rewrite-java-test/src/test/java/org/openrewrite/java/cleanup/RemoveInstanceOfPatternMatchTest.java

```diff
@@ -0,0 +1,880 @@
+                      if (obj instanceof String str && str.isEmpty() || false) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String && ((String) obj).isEmpty() || false) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    void localVariable() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String str || true) {
+                          String str = null;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String || true) {
+                          String str = null;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    @Disabled("Not supported")
+    void negationLocalVariable() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (!(obj instanceof String str)) {
+                          Integer str = null;
+                          System.out.println(str);
+                      } else {
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (!(obj instanceof String)) {
+                          String str = null;
+                          System.out.println(str);
+                      } else {
+                          String str = (String) obj;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    void sameNamedVariables() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String v && v.isEmpty() || obj instanceof Integer v && v.equals(1)) {
```

Comment:
This isn't actually legal Java code. We are working on optionally also validating the before sources, but aren't there quite yet. But here you essentially declare two variables with the same name in the same scope. So I think you can probably just remove this test.

Reply:
I'm not 100% sure, but I think it's legal because two declarations are separated by a logical disjunction. So LHS and RHS of the disjunction forms separate scopes. And Eclipse and IDEA doesn't show syntax errors for such a code.

Reply:
That seems weird. The Javac 17 compiler doesn't accept it and while I can't see any error highlighted in the block string in IDEA, it does show me an error when I try this outside of a string.

Reply:
Yes, you are right. Eclipse doesn't show the error, but IDEA does. I removed the test case.

As Eclipse didn't show any errors I thought it's a correct construction and even found an explanation for myself why it works.